README.md
# go-astm ![build status](https://travis-ci.org/78bit/uuid.svg?branch=master)

Golang library for handling ASTM lis2a2 Procotol

###### Install
`go get github.com/DRK-Blutspende-BaWueHe/go-astm`

## Features
  - Encoding 
    - UTF8 
    - ASCII
    - Windows1250 
    - Windows1251 
    - Windows1252 
    - DOS852 
    - DOS855 
    - DOS866 
  - Timezone conversion on marshal and unmarshal
  - Marshalling and Unmarshalling supported
  - Custom delimiters are recognized in the Header and appplied (defaults are \^&)
  - Supported Types : string, float32, float64, time.Time, enums, int

## Installation

Install the package with the following command.

``` shell
go get github.com/DRK-Blutspende-BaWueHe/go-astm/...
```

## Messaage Structure

Every instrument is a little different. Use the lis2a2-default implmenetation provided with this library as a starting point.

``` go
type CommentedResult struct {
	Result  Result    `astm:"R"`
	Comment []Comment `astm:"C,optional"`
}

type PORC struct {
	Patient         Patient   `astm:"P"`
	Comment         []Comment `astm:"C,optional"`
	Order           Order     `astm:"O"`
	CommentedResult []CommentedResult
}

type DefaultMessage struct {
	Header       Header       `astm:"H"`
	Manufacturer Manufacturer `astm:"M,optional"`
	OrderResults []PORC
	Terminator   Terminator `astm:"L"`
}
```

## Reading ASTM

The following Go code decodes a ASTM provided as a string and stores all its information in the &message.

``` go
var message lis2a2.DefaultMessage

err := lis2a2.Unmarshal([]byte(textdata), &message,
		lis2a2.EncodingUTF8, lis2a2.TimezoneEuropeBerlin)
if err != nil {
  log.Fatal(err)		
}
```

## Writing ASM

Converting an annotated Structure (see above) to an enocded bytestream. 

The bytestream is encoded by-row, lacking the CR code at the end. 

``` go
lines, err := lis2a2.Marshal(msg, lis2a2.EncodingASCII, lis2a2.TimezoneEuropeBerlin, lis2a2.ShortNotation)

// output on screen
for _, line := range lines {
		linestr := string(line)
		fmt.Println(linestr)
}
```
## Message Structure and Annotation

### Optional records
``` go
type SimpleMessage struct  {
	Header       standardlis2a2.Header       `astm:"H"`
	Manufacturer standardlis2a2.Manufacturer `astm:"M,optional"`
	Patient      standardlis2a2.Patient      `astm:"P"`
	Order        standardlis2a2.Order        `astm:"O"`
	Result       standardlis2a2.Result       `astm:"R"`
	Terminator   standardlis2a2.Terminator   `astm:"L"`
}
```

### Nested arrays
``` go
type MessagePORC struct {
	Header       standardlis2a2.Header       `astm:"H"`
	Manufacturer standardlis2a2.Manufacturer `astm:"M,optional"`
	OrderResults []struct {
		Patient         standardlis2a2.Patient `astm:"P"`
		Order           standardlis2a2.Order   `astm:"O"`
		CommentedResult []struct {
			Result  standardlis2a2.Result    `astm:"R"`
			Comment []standardlis2a2.Comment `astm:"C,optional"`
		}
	}
	Terminator standardlis2a2.Terminator `astm:"L"`
}
```

## Custom Record Structure

### Addressing fields 
Often the default is not enough. You can customize any record with annotation. 

#### ... by Field#
``` go
   ...
   Filed string `astm:"3"`  // Select 3rd field, start counting with 1
   ...
```
Example:
``` text
	X|field2|field3|field4|		Result: "field3"
	X|field2^1^2|field3^1^2|field4^5^6|		Result: "field3"	
	X|field2^1^2|field3_1^1_1^2_!\\field3_2^5_2^2_2|field4^6^2|		Result: "field3_1"
```

#### ... by Field#.Component#
``` go
   ...
   Filed string `astm:"3.2"`  // Select 3rd field, 2nd component, start counting with 1
   ...
```
Example:
``` text
	X|field2|field3|field4|		Result: ""	
	X|field2^1^2|field3^1^2|field4^5^6|		Result: "1"	
	X|field2^1^2|field3_1^1_1^2_!\\field3_2^1_2^2_2|field4^1^2|		Result: "1_1"
```
#### ... by Field#.Repeat#.Component#
``` go
   ...
   Filed string `astm:"3.2.2"`  // Select 3rd field, 2nd array index, 2nd component, start counting with 1
   ...
```
Example:
``` text
	X|field2|field3|field4|		Result: ""	
	X|field2^1^2|field3^1^2|field4^5^6|		Result: ""	
	X|field2^1^2|field3_1^1_1^2_!\\field3_2^1_2^2_2|field4^1^2|		Result: "1_2"
```
### Custom Record Format
``` go
type Result struct {
	SequenceNumber                           int       `astm:"2,sequence"`   // sequence generates numbers when value is 0 
	UniversalTestID                          string    `astm:"3.1"`         
	UniversalTestIDName                      string    `astm:"3.2"`         
	UniversalTestIDType                      string    `astm:"3.3"`         
	ManufacturersTestType                    string    `astm:"3.4"`         
	ManufacturersTestName                    string    `astm:"3.5"`         
	ManufacturersTestCode                    string    `astm:"3.6"`         
	TestCode                                 string    `astm:"3.7"`         
	DataMeasurementValue                     string    `astm:"4.1"`         
	InitialMeasurementValue                  string    `astm:"4.2"`         
	MeasurementValueOfDevice                 string    `astm:"4.3"`         
	Units                                    string    `astm:"5"`           
	ReferenceRange                           string    `astm:"6"`           
	ResultAbnormalFlag                       string    `astm:"7"`           
	NatureOfAbnormalTesting                  string    `astm:"8"`           
	ResultStatus                             string    `astm:"9"`           
	DateOfChangeInInstrumentNormativeTesting time.Time `astm:"10,longdate"` 
	OperatorIDPerformed                      string    `astm:"11.1"`        
	OperatorIDVerified                       string    `astm:"11.2"`        
	DateTimeTestStarted                      time.Time `astm:"12,longdate"` 
	DateTimeCompleted                        time.Time `astm:"13,longdate"` 
	IntstrumentIdentification                string    `astm:"14"`          
}
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.9.4] - 2022-06-27

### Fixed

- ignore fields with no annotation

## [0.9.3] - 2022-06-??

### Fixed

## [0.9.2] - 2022-06-??

### Fixed

## [0.9.1] - 2022-06-??

### Fixed

## [0.9.0] - 2022-06-15

### Added

- Base project

### Changed

### Fixed

lib/standardlis2a2/messageformat.go
//
// Standard implementation for the lis2a2 protocol according to standard in every detail,
// will work for most with some tinkering....
//
package standardlis2a2

import "time"

// (see https://samson-rus.com/wp-content/files/LIS2-A2.pdf)
type Header struct {
	Delimiters              string    `astm:"2,delimiter"` // 6.2
	MessageControlID        string    `astm:"3"`           // 6.3
	AccessPassword          string    `astm:"4"`           // 6.4
	SenderNameOrID          string    `astm:"5"`           // 6.5
	SenderStreetAddress     string    `astm:"6"`           // 6.6
	Field6                  string    `astm:"7"`           // 6.7
	SenderTelephone         string    `astm:"8"`           // 6.8
	CharacteristicsOfSender string    `astm:"9"`           // 6.9
	ReceiverID              string    `astm:"10"`          // 6.10
	Comment                 string    `astm:"11"`          // 6.11
	ProcessingID            string    `astm:"12"`          // 6.12
	Version                 string    `astm:"13"`          // 6.13
	DateAndTime             time.Time `astm:"14,longdate"` // 6.14
}

// https://samson-rus.com/wp-content/files/LIS2-A2.pdf
type Patient struct {
	SequenceNumber                     int       `astm:"2,sequence"` // 7.2 (see https://samson-rus.com/wp-content/files/LIS2-A2.pdf)
	PracticeAssignedPatientID          string    `astm:"3"`          // 7.3
	LabAssignedPatientID               string    `astm:"4"`          // 7.4
	ID3                                string    `astm:"5"`          // 7.5
	LastName                           string    `astm:"6.1"`        // 7.6.1
	FirstName                          string    `astm:"6.2"`        // 7.6.2
	MothersMaidenName                  string    `astm:"7"`          // 7.7
	DOB                                time.Time `astm:"8"`          // 7.8
	Gender                             string    `astm:"9"`          // 7.9
	Race                               string    `astm:"10"`         // 7.10
	Address                            string    `astm:"11"`         // 7.11
	F12                                string    `astm:"12"`         // 7.12
	Telephone                          string    `astm:"13"`         // 7.13
	AttendingPhysicianID               string    `astm:"14"`         // 7.14
	SpecialField1                      string    `astm:"15"`         // 7.15
	SpecialField2                      string    `astm:"16"`         // 7.16
	Height                             string    `astm:"17"`         // 7.17
	Weight                             string    `astm:"18"`         // 7.18
	SuspectedDiagnosis                 string    `astm:"19"`         // 7.19
	ActiveMedication                   string    `astm:"20"`         // 7.20
	Diet                               string    `astm:"21"`         // 7.21
	PracticeField1                     string    `astm:"22"`         // 7.22
	PracticeField2                     string    `astm:"23"`         // 7.23
	AdmissionAndDischargeDates         string    `astm:"24"`         // 7.24
	AdmissionStatus                    string    `astm:"25"`         // 7.25
	Location                           string    `astm:"26"`         // 7.26
	NatureOfAlternativeDiagnosticCodes string    `astm:"27"`         // 7.27
	AlternativeDiagnosticCodes         string    `astm:"28"`         // 7.28
	Religion                           string    `astm:"29"`         // 7.29
	MaritalStatus                      string    `astm:"30"`         // 7.30
	IsolationStatus                    string    `astm:"31"`         // 7.31
	Language                           string    `astm:"32"`         // 7.32
	HospitalService                    string    `astm:"33"`         // 7.33
	HospitalInstitution                string    `astm:"34"`         // 7.34
	DosageCategory                     string    `astm:"35"`         // 7.35
}

// https://samson-rus.com/wp-content/files/LIS2-A2.pdf
type Order struct {
	SequenceNumber               int       `astm:"2,sequence"`  // 8.4.2 (see https://samson-rus.com/wp-content/files/LIS2-A2.pdf)
	SpecimenID                   string    `astm:"3"`           // 8.4.3
	InstrumentSpecimenID         string    `astm:"4"`           // 8.4.4
	UniversalTestID              string    `astm:"5"`           // 8.4.5
	Priority                     string    `astm:"6"`           // 8.4.6
	RequestedOrderDateTime       time.Time `astm:"7,longdate"`  // 8.4.7
	SpecimenCollectionDateTime   time.Time `astm:"8,longdate"`  // 8.4.8
	CollectionEndTime            time.Time `astm:"9,longdate"`  // 8.4.9
	CollectionVolume             string    `astm:"10"`          // 8.4.10
	CollectionID                 string    `astm:"11"`          // 8.4.11
	ActionCode                   string    `astm:"12"`          // 8.4.12
	DangerCode                   string    `astm:"13"`          // 8.4.13
	RelevantClinicalInformation  string    `astm:"14"`          // 8.4.14
	DateTimeSpecimenReceived     string    `astm:"15"`          // 8.4.15
	SpecimenDescriptor           string    `astm:"16"`          // 8.4.16
	OrderingPhysician            string    `astm:"17"`          // 8.4.17
	PhysicianTelephone           string    `astm:"18"`          // 8.4.18
	UserField1                   string    `astm:"19"`          // 8.4.19
	UserField2                   string    `astm:"20"`          // 8.4.20
	LaboratoryField1             string    `astm:"21"`          // 8.4.21
	LaboratoryField2             string    `astm:"22"`          // 8.4.22
	DateTimeResultsReported      time.Time `astm:"23,longdate"` // 8.4.23
	InstrumentCharge             string    `astm:"24"`          // 8.4.24
	InstrumentSectionID          string    `astm:"25"`          // 8.4.25
	ReportType                   string    `astm:"26"`          // 8.4.26
	Reserved                     string    `astm:"27"`          // 8.4.27
	LocationOfSpecimenCollection string    `astm:"28"`          // 8.4.28
	NosocomialInfectionFlag      string    `astm:"29"`          // 8.4.29
	SpecimenService              string    `astm:"30"`          // 8.4.30
	SpecimenInstitution          string    `astm:"31"`          // 8.4.31
}

// https://samson-rus.com/wp-content/files/LIS2-A2.pdf
type Result struct {
	SequenceNumber                           int       `astm:"2,sequence"`  // 9.2 https://samson-rus.com/wp-content/files/LIS2-A2.pdf
	UniversalTestID                          string    `astm:"3.1"`         // 9.3
	UniversalTestIDName                      string    `astm:"3.2"`         // 9.3
	UniversalTestIDType                      string    `astm:"3.3"`         // 9.3
	ManufacturersTestType                    string    `astm:"3.4"`         // 9.3
	ManufacturersTestName                    string    `astm:"3.5"`         // 9.3
	ManufacturersTestCode                    string    `astm:"3.6"`         // 9.3
	TestCode                                 string    `astm:"3.7"`         // 9.3
	DataMeasurementValue                     string    `astm:"4.1"`         // 9.4
	InitialMeasurementValue                  string    `astm:"4.2"`         // 9.4
	MeasurementValueOfDevice                 string    `astm:"4.3"`         // 9.4
	Units                                    string    `astm:"5"`           // 9.5
	ReferenceRange                           string    `astm:"6"`           // 9.6
	ResultAbnormalFlag                       string    `astm:"7"`           // 9.7
	NatureOfAbnormalTesting                  string    `astm:"8"`           // 9.8
	ResultStatus                             string    `astm:"9"`           // 9.9
	DateOfChangeInInstrumentNormativeTesting time.Time `astm:"10,longdate"` // 9.10
	OperatorIDPerformed                      string    `astm:"11.1"`        // 9.11
	OperatorIDVerified                       string    `astm:"11.2"`        // 9.11
	DateTimeTestStarted                      time.Time `astm:"12,longdate"` // 9.12
	DateTimeCompleted                        time.Time `astm:"13,longdate"` // 9.13
	IntstrumentIdentification                string    `astm:"14"`          // 9.14
}

// https://samson-rus.com/wp-content/files/LIS2-A2.pdf
type Comment struct {
	SequenceNumber int    `astm:"2,sequence"` // 10.2 (see https://samson-rus.com/wp-content/files/LIS2-A2.pdf)
	CommentSource  string `astm:"3"`          // 10.3
	CommentText    string `astm:"4"`          // 10.4
	CommentType    string `astm:"5"`          // 10.5
}

// Lis2RequestInformation - RequestInformation
// https://samson-rus.com/wp-content/files/LIS2-A2.pdf
type RequestInformation struct {
	SequenceNumber                  int    `astm:"2,sequence"` // 11.2 (see https://samson-rus.com/wp-content/files/LIS2-A2.pdf)
	StartingRangeIDNumber           string `astm:"3"`          // 11.3
	EndingRangeIDNumber             string `astm:"4"`          // 11.4
	UniversalTestID                 string `astm:"5"`          // 11.5
	NatureOfRequestTimeLimits       string `astm:"6"`          // 11.6
	BeginningRequestResultsDateTime string `astm:"7"`          // 11.7
	EndingRequestResultsDateTime    string `astm:"8"`          // 11.8
	RequestingPhysicianName         string `astm:"9"`          // 11.9
	RequestingPhysicianTelephone    string `astm:"10"`         // 11.10
	UserField1                      string `astm:"11"`         // 11.11
	UserField2                      string `astm:"12"`         // 11.12
	RequestInformationStatus        string `astm:"13"`         // 11.13
}

// Lis2Terminator - Terminator Record (Hasta la vista....)
// https://samson-rus.com/wp-content/files/LIS2-A2.pdf
type Terminator struct {
	SequenceNumber int    `astm:"2,sequence"` // 12.2 (see https://samson-rus.com/wp-content/files/LIS2-A2.pdf)
	TerminatorCode string `astm:"3"`          // 12.3
}

// Lis2Manufacturer -Manufacturer Record
// https://samson-rus.com/wp-content/files/LIS2-A2.pdf
type Manufacturer struct {
	SequenceNumber string `astm:"2,sequence"` // 14.2 (see https://samson-rus.com/wp-content/files/LIS2-A2.pdf)
	F2             string `astm:"3"`          // 14.3
	F3             string `astm:"4"`          // 14.4
	F4             string `astm:"5"`          // 14.5
	F5             string `astm:"6"`          // 14.6
	F6             string `astm:"7"`          // 14.7
	F7             string `astm:"8"`          // 14.8
	F8             string `astm:"9"`          // 14.9
	F9             string `astm:"10"`         // 14.10
	F10            string `astm:"11"`         // 14.11
	F11            string `astm:"12"`         // 14.12
	F12            string `astm:"13"`         // 14.13
	F13            string `astm:"14"`         // 14.14
}

type CommentedResult struct {
	Result  Result    `astm:"R"`
	Comment []Comment `astm:"C,optional"`
}

type PORC struct {
	Patient         Patient   `astm:"P"`
	Comment         []Comment `astm:"C,optional"`
	Order           Order     `astm:"O"`
	CommentedResult []CommentedResult
}

// https://samson-rus.com/wp-content/files/LIS2-A2.pdf Page 30 : Logial Structure of Message
type DefaultMessage struct {
	Header       Header       `astm:"H"`
	Manufacturer Manufacturer `astm:"M,optional"`
	OrderResults []PORC
	Terminator   Terminator `astm:"L"`
}
